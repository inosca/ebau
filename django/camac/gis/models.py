from django.db import models
from django.utils.translation import gettext_lazy as _


class GISConfig(models.Model):
    CLIENT_SOGIS = "camac.gis.clients.sogis.SoGisClient"
    CLIENT_PARAM = "camac.gis.clients.param.ParamClient"

    CLIENT_CHOICES = [
        (CLIENT_SOGIS, _("GIS Canton Solothurn")),
        (CLIENT_PARAM, _("Parameter")),
    ]

    slug = models.SlugField(primary_key=True)
    client = models.CharField(max_length=255, choices=CLIENT_CHOICES)
    config = models.JSONField(default=dict)
    disabled = models.BooleanField(default=False)
