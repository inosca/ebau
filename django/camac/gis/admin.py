from adminsortable2.admin import SortableAdminMixin
from django.contrib.admin import ModelAdmin, register
from django.db.models import JSONField
from django_json_widget.widgets import JSONEditorWidget

from camac.admin import EbauAdminMixin
from camac.gis.models import GISDataSource


@register(GISDataSource)
class GISDataSourceAdmin(EbauAdminMixin, SortableAdminMixin, ModelAdmin):
    list_display = ["description", "client", "disabled"]
    list_per_page = 50
    search_fields = ["description", "client"]
    formfield_overrides = {
        JSONField: {"widget": JSONEditorWidget},
    }
