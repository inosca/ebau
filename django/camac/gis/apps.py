from django.apps import AppConfig


class GisConfig(AppConfig):
    default_auto_field = "django.db.models.BigAutoField"
    name = "camac.gis"
