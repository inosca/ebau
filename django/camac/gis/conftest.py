import pytest
from syrupy.filters import props


@pytest.fixture
def mock_municipalities(mocker):
    def mock(names=[]):
        mocker.patch(
            "camac.caluma.extensions.data_sources.Municipalities.get_data",
            return_value=[[i + 1, {"de": name}] for i, name in enumerate(names)],
        )

    return mock


@pytest.fixture
def gis_snapshot(snapshot):
    return snapshot(exclude=props("cache"))
