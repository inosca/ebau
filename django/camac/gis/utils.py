from enum import Enum
from typing import Any, List, Union

from django.utils.translation import gettext as _


class MergeStrategy(Enum):
    MERGE_FIRST = "merge_first"

    # actually not merge anything at all, give full control over return value
    # to each client (later clients can change output of previous clients)
    OVERRIDE = "override"


def get_bbox(x: Union[float, str], y: Union[float, str], buffer: int = 0) -> str:
    delta = 0

    if buffer:
        delta = buffer / 2

    try:
        x = float(x)
        y = float(y)
    except ValueError:
        raise ValueError(_("Coordinates must be floats"))

    return ",".join(map(str, [x - delta, y - delta, x + delta, y + delta]))


def to_query(query_params: dict) -> str:
    return "&".join(
        [f"{key}={value}" for key, value in query_params.items() if value is not None]
    )


def cast(value: Any, type: str) -> Any:
    if value is None:
        return None

    try:
        if type == "float":
            return float(value)
        elif type == "integer":
            return int(value)
        elif type == "string":
            return str(value)
    except ValueError:
        return None

    return value


def concat_values(*parts: Any) -> str:
    """Clean, deduplicate and concatenate values.

    This function cleans out empty values, deduplicates, and concatenates them
    using a comma as separator. If there is no meaningful value, it returns none
    and if there is only one value it returns it without converting it to a
    string.
    """

    # Filter out empty values
    non_empty_parts = [part for part in parts if part not in [None, ""]]

    # Return None if there are only empty values
    if len(non_empty_parts) == 0:
        return None
    # Return single value without joining to preserve the data type
    elif len(non_empty_parts) == 1:
        value = non_empty_parts[0]

        if isinstance(value, str):
            value = value.strip()

        return value

    clean_parts = [str(part).strip() for part in non_empty_parts]
    deduped_parts = sorted(set(clean_parts), key=clean_parts.index)

    return ", ".join(deduped_parts).strip()


def merge_table(
    existing_rows: List[dict],
    new_row: dict,
    merge_strategy: MergeStrategy,
) -> List[dict]:
    if merge_strategy == MergeStrategy.MERGE_FIRST:
        merge_data(existing_rows[0], new_row, merge_strategy)

    return existing_rows


def merge_data(data: dict, new_data: dict, merge_strategy: MergeStrategy) -> dict:
    for key, value in new_data.items():
        if key in data:
            if isinstance(data[key], list):
                for row in value:
                    value = merge_table(data[key], row, merge_strategy)
            elif merge_strategy == MergeStrategy.MERGE_FIRST:
                # If a previous data source already returned a value for a
                # certain question we concat the new and old value
                value = concat_values(data[key], value)

        data[key] = value
