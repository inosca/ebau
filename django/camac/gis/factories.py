from factory import Faker, fuzzy
from factory.django import DjangoModelFactory

from camac.gis import models


class GISDataSourceFactory(DjangoModelFactory):
    description = Faker("text")
    client = fuzzy.FuzzyChoice(dict(models.GISDataSource.CLIENT_CHOICES).keys())
    config = {}
    disabled = False

    class Meta:
        model = models.GISDataSource
