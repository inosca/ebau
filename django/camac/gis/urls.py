from django.urls import re_path

from camac.gis import views

urlpatterns = [
    re_path(
        r"^data/(?:(?P<task_id>[\w-]+)/)?$",
        views.GISDataView.as_view(),
        name="gis-data",
    ),
    re_path(
        r"^apply$",
        views.GISApplyView.as_view(),
        name="gis-apply",
    ),
]
