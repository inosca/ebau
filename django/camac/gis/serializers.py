from dataclasses import dataclass
from typing import List

from caluma.caluma_form.api import save_answer
from caluma.caluma_form.models import Document, Question
from django.core.cache import cache
from django.utils.translation import gettext as _
from rest_framework import serializers
from rest_framework.exceptions import PermissionDenied, ValidationError

from camac.caluma.extensions.permissions import CustomPermission
from camac.instance.models import Instance


@dataclass
class GISApplyResponse:
    pk: None
    instance: Instance
    questions: List[str]


class GISApplySerializer(serializers.Serializer):
    instance = serializers.PrimaryKeyRelatedField(queryset=Instance.objects)
    cache = serializers.UUIDField(write_only=True)
    questions = serializers.ListField(child=serializers.CharField(), read_only=True)

    def validate_cache(self, value):
        if not cache.has_key(value):
            raise ValidationError(
                _("Cache for key %(key)s does not exist") % {"key": value}
            )

        return value

    def _write_row(self, form_slug, row_data):
        row_document = Document.objects.create(form_id=form_slug)

        for question_slug, value in row_data.items():
            self._write_answer(row_document, question_slug, value)

        return str(row_document.pk)

    def _write_answer(self, document, question_slug, value):
        try:
            question = Question.objects.get(pk=question_slug)
        except Question.DoesNotExist:  # pragma: no cover
            return False

        answer_value = value["value"]

        if question.type == Question.TYPE_TABLE:
            answer_value = [
                self._write_row(value["form"], row) for row in value["value"]
            ]
        elif question.type == Question.TYPE_MULTIPLE_CHOICE:
            answer_value = [opt["value"] for opt in answer_value]

        save_answer(
            question=question,
            document=document,
            user=self.context["request"].caluma_info.context.user,
            value=int(answer_value)
            if question.type == "integer"
            else float(answer_value)
            if question.type == "float"
            else answer_value,
            meta={"gis-value": answer_value},
        )

        return True

    def create(self, validated_data):
        data = cache.get(validated_data["cache"])
        document = validated_data["instance"].case.document
        written_questions = set()

        if not CustomPermission(self.context["request"]).has_camac_edit_permission(
            document, self.context["request"].caluma_info
        ):
            raise PermissionDenied()

        for question_slug, value in data.items():
            success = self._write_answer(document, question_slug, value)

            if success:
                written_questions.add(question_slug)

        return GISApplyResponse(
            pk=None,
            instance=validated_data["instance"],
            questions=written_questions,
        )
