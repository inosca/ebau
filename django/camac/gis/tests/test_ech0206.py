import json

import pytest
from caluma.caluma_form.models import Question
from django.urls import reverse
from rest_framework import status

from camac.gis.models import GISDataSource


@pytest.fixture
def ech0206_data_sources(
    question_factory, question_option_factory, option_factory, settings
):
    gis_questions = [
        ("egid-nr", Question.TYPE_TEXT),
        (
            "energie-waermequelle-heizung",
            Question.TYPE_CHOICE,
            [
                "7500",
                "7501",
                "7510",
                "7511",
                "7512",
                "7513",
                "7520",
                "7530",
                "7540",
                "7541",
                "7542",
                "7550",
                "7560",
                "7570",
                "7580",
                "7581",
                "7582",
                "7598",
                "7599",
                "noch-nicht-festgelegt",
            ],
        ),
        (
            "energie-waermequelle-warmwasser",
            Question.TYPE_CHOICE,
            [
                "7500",
                "7501",
                "7510",
                "7511",
                "7512",
                "7513",
                "7520",
                "7530",
                "7540",
                "7541",
                "7542",
                "7550",
                "7560",
                "7570",
                "7580",
                "7581",
                "7582",
                "7598",
                "7599",
                "noch-nicht-festgelegt",
            ],
        ),
        (
            "waermeerzeuger-warmwasser",
            Question.TYPE_CHOICE,
            [
                "7600",
                "7610",
                "7620",
                "7630",
                "7632",
                "7634",
                "7640",
                "7650",
                "7651",
                "7660",
                "7699",
            ],
        ),
        (
            "waermeerzeuger-heizung",
            Question.TYPE_CHOICE,
            [
                "7400",
                "7410",
                "7420",
                "7430",
                "7431",
                "7432",
                "7433",
                "7434",
                "7435",
                "7436",
                "7440",
                "7441",
                "7450",
                "7451",
                "7460",
                "7461",
                "7499",
                "noch-nicht-festgelegt",
            ],
        ),
    ]

    for config in gis_questions:
        slug = config[0]
        question_type = config[1]
        q = question_factory(slug=slug, type=question_type, label=slug)
        if len(config) == 3:
            for option in config[2]:
                question_option_factory(
                    question=q,
                    option=option_factory(slug=f"{slug}-{option}", label=option),
                )

    return GISDataSource.objects.all()


@pytest.fixture
def ech0206__config(gis_data_source_factory, question_factory):
    gis_data_source_factory(
        client=GISDataSource.CLIENT_KT_GR,
        config=[
            {
                "identifier": "gebaeudeadressen_av",
                "properties": [
                    {"question": "egid-nr", "propertyName": "gwr_egid"},
                ],
            },
        ],
    )
    gis_data_source_factory(
        client=GISDataSource.CLIENT_ECH_0206,
        config=[
            {
                "properties": [
                    {
                        "question": "energie-waermequelle-heizung",
                        "propertyName": "thermotechnicalDeviceForHeating1.energySourceHeating",
                    },
                    {
                        "question": "weitere-energie-waermequelle-heizung",
                        "propertyName": "thermotechnicalDeviceForHeating2.energySourceHeating",
                    },
                    {
                        "question": "waermeerzeuger-heizung",
                        "propertyName": "thermotechnicalDeviceForHeating1.heatGeneratorHeating",
                    },
                    {
                        "question": "weitere-waermeerzeuger-heizung",
                        "propertyName": "thermotechnicalDeviceForHeating2.heatGeneratorHeating",
                    },
                    {
                        "question": "energie-waermequelle-warmwasser",
                        "propertyName": "thermotechnicalDeviceForWarmWater1.energySourceHeating",
                    },
                    {
                        "question": "weitere-energie-waermequelle-warmwasser",
                        "propertyName": "thermotechnicalDeviceForWarmWater2.energySourceHeating",
                    },
                    {
                        "question": "waermeerzeuger-warmwasser",
                        "propertyName": "thermotechnicalDeviceForWarmWater1.heatGeneratorHotWater",
                    },
                    {
                        "question": "weitere-waermeerzeuger-warmwasser",
                        "propertyName": "thermotechnicalDeviceForWarmWater2.heatGeneratorHotWater",
                    },
                ]
            },
            {
                "properties": [
                    {
                        "question": "amtliche-gebaeudenummer",
                        "propertyName": "officialBuildingNo",
                    }
                ]
            },
            {
                "properties": [
                    {"question": "gwr-gebaeudevolumen", "propertyName": "volume"}
                ]
            },
            {
                "properties": [
                    {
                        "question": "energiebezugsflaeche",
                        "propertyName": "energyRelevantSurface",
                    }
                ]
            },
            {
                "properties": [
                    {
                        "hidden": True,
                        "mapper": "heat_surface",
                        "question": "werden-flaechen-beheizt",
                    }
                ]
            },
            {
                "properties": [
                    {
                        "hidden": True,
                        "mapper": "warm_water_connection",
                        "question": "ist-ein-warmwasseranschluss-geplant",
                    }
                ]
            },
        ],
    )


@pytest.mark.parametrize(
    "query_x,query_y",
    [
        (
            2759664.9038492967,
            1191395.5015286694,
        ),  # multiple egid numbers with multiple properties
        (
            2759647.4234941704,
            1191526.8007143869,
        ),  # egid nr with Energiebezugsfläche
        (
            2759652.9068137,
            1191423.1268269215,
        ),  # no egid number
        (
            2759746.044642407,
            1191415.6188414234,
        ),  # single egid number
        (
            2759772.4989033034,
            1191611.9958049427,
        ),  # Energie-/Wärmequelle Heizung= Unbestimmt & official building = NN
        (
            2760073.7106244904,
            1191456.8678512666,
        ),  # Werden Flächen beheizt? --> Nein  and Ist ein Warmwasseranschluss geplant? --> Nein
        (
            2759626.1504406286,
            1191908.3321774867,
        ),  # Ist ein Warmwasseranschluss geplant? --> Nein
    ],
)
@pytest.mark.vcr()
def test_ech0206_client(
    db,
    admin_client,
    snapshot,
    vcr_config,
    ech0206__config,
    query_x,
    query_y,
    ech0206_data_sources,
):
    query = {"markers": [{"x": query_x, "y": query_y}], "geometry": "POINT"}
    response = admin_client.get(
        reverse("gis-data"),
        data={
            "query": json.dumps(query),
            "form": "baugesuch",
        },
    )

    assert response.status_code == status.HTTP_200_OK
    snapshot.assert_match(response.json())
