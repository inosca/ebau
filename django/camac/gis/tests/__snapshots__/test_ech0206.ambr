# serializer version: 1
# name: test_ech0206_client[2759626.1504406286-1191908.3321774867]
  dict({
    'data': dict({
      'folgeplanung': dict({
        'hidden': False,
        'label': None,
        'value': 'Quartierplan Roter Turm (hinweisend)',
      }),
      'gebaeude-und-anlagen': dict({
        'hidden': False,
        'label': None,
        'value': list([
          dict({
            'amtliche-gebaeudenummer': dict({
              'label': None,
              'value': '4-242B',
            }),
            'egid-nr': dict({
              'label': 'egid-nr',
              'value': '400017774',
            }),
            'energie-waermequelle-heizung': dict({
              'displayValue': '7520',
              'label': 'energie-waermequelle-heizung',
              'value': 'energie-waermequelle-heizung-7520',
            }),
            'energie-waermequelle-warmwasser': dict({
              'displayValue': '7500',
              'label': 'energie-waermequelle-warmwasser',
              'value': 'energie-waermequelle-warmwasser-7500',
            }),
            'ist-ein-warmwasseranschluss-geplant': dict({
              'label': None,
              'value': 'nein',
            }),
            'waermeerzeuger-heizung': dict({
              'displayValue': '7430',
              'label': 'waermeerzeuger-heizung',
              'value': 'waermeerzeuger-heizung-7430',
            }),
            'waermeerzeuger-warmwasser': dict({
              'displayValue': '7600',
              'label': 'waermeerzeuger-warmwasser',
              'value': 'waermeerzeuger-warmwasser-7600',
            }),
            'weitere-energie-waermequelle-heizung': dict({
              'label': None,
              'value': '7520',
            }),
            'weitere-energie-waermequelle-warmwasser': dict({
              'label': None,
              'value': '7500',
            }),
            'weitere-waermeerzeuger-warmwasser': dict({
              'label': None,
              'value': '7600',
            }),
            'werden-flaechen-beheizt': dict({
              'label': None,
              'value': 'ja',
            }),
          }),
        ]),
      }),
      'genereller-erschliessungsplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Parkierung Gebiete A',
      }),
      'parzelle': dict({
        'hidden': False,
        'label': None,
        'value': list([
          dict({
            'e-grid-nr': dict({
              'label': None,
              'value': 'CH916878867738',
            }),
            'parzellennummer': dict({
              'label': None,
              'value': '129',
            }),
          }),
          dict({
            'baurecht-nummer': dict({
              'label': None,
              'value': '9962',
            }),
            'e-grid-nr': dict({
              'label': None,
              'value': 'CH438677685533',
            }),
          }),
          dict({
            'baurecht-nummer': dict({
              'label': None,
              'value': '9066',
            }),
            'e-grid-nr': dict({
              'label': None,
              'value': 'CH616845867749',
            }),
          }),
        ]),
      }),
      'zonenplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Wohnzone 4, Keine Gefahrenzone',
      }),
    }),
  })
# ---
# name: test_ech0206_client[2759647.4234941704-1191526.8007143869]
  dict({
    'data': dict({
      'gebaeude-und-anlagen': dict({
        'hidden': False,
        'label': None,
        'value': list([
          dict({
            'amtliche-gebaeudenummer': dict({
              'label': None,
              'value': '4-116',
            }),
            'egid-nr': dict({
              'label': 'egid-nr',
              'value': '1738776',
            }),
            'energie-waermequelle-heizung': dict({
              'displayValue': '7501',
              'label': 'energie-waermequelle-heizung',
              'value': 'energie-waermequelle-heizung-7501',
            }),
            'energie-waermequelle-warmwasser': dict({
              'displayValue': '7501',
              'label': 'energie-waermequelle-warmwasser',
              'value': 'energie-waermequelle-warmwasser-7501',
            }),
            'energiebezugsflaeche': dict({
              'label': None,
              'value': '226',
            }),
            'ist-ein-warmwasseranschluss-geplant': dict({
              'label': None,
              'value': 'ja',
            }),
            'waermeerzeuger-heizung': dict({
              'displayValue': '7410',
              'label': 'waermeerzeuger-heizung',
              'value': 'waermeerzeuger-heizung-7410',
            }),
            'waermeerzeuger-warmwasser': dict({
              'displayValue': '7610',
              'label': 'waermeerzeuger-warmwasser',
              'value': 'waermeerzeuger-warmwasser-7610',
            }),
            'weitere-energie-waermequelle-heizung': dict({
              'label': None,
              'value': '7501',
            }),
            'weitere-energie-waermequelle-warmwasser': dict({
              'label': None,
              'value': '7500',
            }),
            'weitere-waermeerzeuger-heizung': dict({
              'label': None,
              'value': '7400',
            }),
            'weitere-waermeerzeuger-warmwasser': dict({
              'label': None,
              'value': '7600',
            }),
            'werden-flaechen-beheizt': dict({
              'label': None,
              'value': 'ja',
            }),
          }),
          dict({
            'amtliche-gebaeudenummer': dict({
              'label': None,
              'value': '4-117',
            }),
            'egid-nr': dict({
              'label': 'egid-nr',
              'value': '1738778',
            }),
            'energie-waermequelle-heizung': dict({
              'displayValue': '7530',
              'label': 'energie-waermequelle-heizung',
              'value': 'energie-waermequelle-heizung-7530',
            }),
            'energie-waermequelle-warmwasser': dict({
              'displayValue': '7530',
              'label': 'energie-waermequelle-warmwasser',
              'value': 'energie-waermequelle-warmwasser-7530',
            }),
            'ist-ein-warmwasseranschluss-geplant': dict({
              'label': None,
              'value': 'ja',
            }),
            'waermeerzeuger-heizung': dict({
              'displayValue': '7430',
              'label': 'waermeerzeuger-heizung',
              'value': 'waermeerzeuger-heizung-7430',
            }),
            'waermeerzeuger-warmwasser': dict({
              'displayValue': '7630',
              'label': 'waermeerzeuger-warmwasser',
              'value': 'waermeerzeuger-warmwasser-7630',
            }),
            'weitere-energie-waermequelle-heizung': dict({
              'label': None,
              'value': '7530',
            }),
            'weitere-energie-waermequelle-warmwasser': dict({
              'label': None,
              'value': '7500',
            }),
            'weitere-waermeerzeuger-warmwasser': dict({
              'label': None,
              'value': '7600',
            }),
            'werden-flaechen-beheizt': dict({
              'label': None,
              'value': 'ja',
            }),
          }),
          dict({
            'amtliche-gebaeudenummer': dict({
              'label': None,
              'value': '4-115B',
            }),
            'egid-nr': dict({
              'label': 'egid-nr',
              'value': '1738793',
            }),
            'energie-waermequelle-heizung': dict({
              'displayValue': '7530',
              'label': 'energie-waermequelle-heizung',
              'value': 'energie-waermequelle-heizung-7530',
            }),
            'energie-waermequelle-warmwasser': dict({
              'displayValue': '7530',
              'label': 'energie-waermequelle-warmwasser',
              'value': 'energie-waermequelle-warmwasser-7530',
            }),
            'ist-ein-warmwasseranschluss-geplant': dict({
              'label': None,
              'value': 'ja',
            }),
            'waermeerzeuger-heizung': dict({
              'displayValue': '7430',
              'label': 'waermeerzeuger-heizung',
              'value': 'waermeerzeuger-heizung-7430',
            }),
            'waermeerzeuger-warmwasser': dict({
              'displayValue': '7630',
              'label': 'waermeerzeuger-warmwasser',
              'value': 'waermeerzeuger-warmwasser-7630',
            }),
            'weitere-energie-waermequelle-heizung': dict({
              'label': None,
              'value': '7530',
            }),
            'weitere-energie-waermequelle-warmwasser': dict({
              'label': None,
              'value': '7500',
            }),
            'weitere-waermeerzeuger-warmwasser': dict({
              'label': None,
              'value': '7600',
            }),
            'werden-flaechen-beheizt': dict({
              'label': None,
              'value': 'ja',
            }),
          }),
        ]),
      }),
      'genereller-erschliessungsplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Parkierung Gebiete A',
      }),
      'genereller-gestaltungsplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Erhaltungsbereiche Gebiete mit besonderer Wohnqualit√§t',
      }),
      'parzelle': dict({
        'hidden': False,
        'label': None,
        'value': list([
          dict({
            'e-grid-nr': dict({
              'label': None,
              'value': 'CH287786687730',
            }),
            'parzellennummer': dict({
              'label': None,
              'value': '54',
            }),
          }),
        ]),
      }),
      'zonenplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Wohnzone 3, Keine Gefahrenzone',
      }),
    }),
  })
# ---
# name: test_ech0206_client[2759652.9068137-1191423.1268269215]
  dict({
    'data': dict({
      'folgeplanung': dict({
        'hidden': False,
        'label': None,
        'value': 'Baulinie allgemein',
      }),
      'genereller-erschliessungsplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Fuss- / Spazierweg, Parkierung Gebiete A',
      }),
      'genereller-gestaltungsplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Erhaltungsbereiche Gebiete mit besonderer Wohnqualit√§t',
      }),
      'parzelle': dict({
        'hidden': False,
        'label': None,
        'value': list([
          dict({
            'e-grid-nr': dict({
              'label': None,
              'value': 'CH438677680484',
            }),
            'parzellennummer': dict({
              'label': None,
              'value': '5084',
            }),
          }),
        ]),
      }),
      'zonenplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Wohnzone 3, Keine Gefahrenzone',
      }),
    }),
  })
# ---
# name: test_ech0206_client[2759664.9038492967-1191395.5015286694]
  dict({
    'data': dict({
      'folgeplanung': dict({
        'hidden': False,
        'label': None,
        'value': 'Baulinie allgemein',
      }),
      'gebaeude-und-anlagen': dict({
        'hidden': False,
        'label': None,
        'value': list([
          dict({
            'amtliche-gebaeudenummer': dict({
              'label': None,
              'value': '4-47',
            }),
            'egid-nr': dict({
              'label': 'egid-nr',
              'value': '1738808',
            }),
            'energie-waermequelle-heizung': dict({
              'displayValue': '7520',
              'label': 'energie-waermequelle-heizung',
              'value': 'energie-waermequelle-heizung-7520',
            }),
            'energie-waermequelle-warmwasser': dict({
              'displayValue': '7520',
              'label': 'energie-waermequelle-warmwasser',
              'value': 'energie-waermequelle-warmwasser-7520',
            }),
            'ist-ein-warmwasseranschluss-geplant': dict({
              'label': None,
              'value': 'ja',
            }),
            'waermeerzeuger-heizung': dict({
              'displayValue': '7430',
              'label': 'waermeerzeuger-heizung',
              'value': 'waermeerzeuger-heizung-7430',
            }),
            'waermeerzeuger-warmwasser': dict({
              'displayValue': '7630',
              'label': 'waermeerzeuger-warmwasser',
              'value': 'waermeerzeuger-warmwasser-7630',
            }),
            'weitere-energie-waermequelle-heizung': dict({
              'label': None,
              'value': '7520',
            }),
            'weitere-energie-waermequelle-warmwasser': dict({
              'label': None,
              'value': '7500',
            }),
            'weitere-waermeerzeuger-warmwasser': dict({
              'label': None,
              'value': '7600',
            }),
            'werden-flaechen-beheizt': dict({
              'label': None,
              'value': 'ja',
            }),
          }),
          dict({
            'amtliche-gebaeudenummer': dict({
              'label': None,
              'value': '4-46',
            }),
            'egid-nr': dict({
              'label': 'egid-nr',
              'value': '1738809',
            }),
            'energie-waermequelle-heizung': dict({
              'displayValue': '7520',
              'label': 'energie-waermequelle-heizung',
              'value': 'energie-waermequelle-heizung-7520',
            }),
            'energie-waermequelle-warmwasser': dict({
              'displayValue': '7520',
              'label': 'energie-waermequelle-warmwasser',
              'value': 'energie-waermequelle-warmwasser-7520',
            }),
            'ist-ein-warmwasseranschluss-geplant': dict({
              'label': None,
              'value': 'ja',
            }),
            'waermeerzeuger-heizung': dict({
              'displayValue': '7430',
              'label': 'waermeerzeuger-heizung',
              'value': 'waermeerzeuger-heizung-7430',
            }),
            'waermeerzeuger-warmwasser': dict({
              'displayValue': '7630',
              'label': 'waermeerzeuger-warmwasser',
              'value': 'waermeerzeuger-warmwasser-7630',
            }),
            'weitere-energie-waermequelle-heizung': dict({
              'label': None,
              'value': '7520',
            }),
            'weitere-energie-waermequelle-warmwasser': dict({
              'label': None,
              'value': '7500',
            }),
            'weitere-waermeerzeuger-warmwasser': dict({
              'label': None,
              'value': '7600',
            }),
            'werden-flaechen-beheizt': dict({
              'label': None,
              'value': 'ja',
            }),
          }),
        ]),
      }),
      'genereller-erschliessungsplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Parkierung Gebiete A',
      }),
      'genereller-gestaltungsplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Bauten und Anlagen erhaltenswert, Erhaltungsbereiche Gebiete mit besonderer Wohnqualit√§t',
      }),
      'parzelle': dict({
        'hidden': False,
        'label': None,
        'value': list([
          dict({
            'e-grid-nr': dict({
              'label': None,
              'value': 'CH127777866804',
            }),
            'parzellennummer': dict({
              'label': None,
              'value': '42',
            }),
          }),
        ]),
      }),
      'zonenplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Wohnzone 3, Keine Gefahrenzone',
      }),
    }),
  })
# ---
# name: test_ech0206_client[2759746.044642407-1191415.6188414234]
  dict({
    'data': dict({
      'folgeplanung': dict({
        'hidden': False,
        'label': None,
        'value': 'Baulinie allgemein',
      }),
      'gebaeude-und-anlagen': dict({
        'hidden': False,
        'label': None,
        'value': list([
          dict({
            'amtliche-gebaeudenummer': dict({
              'label': None,
              'value': '4-50',
            }),
            'egid-nr': dict({
              'label': 'egid-nr',
              'value': '9001212',
            }),
            'energie-waermequelle-heizung': dict({
              'displayValue': '7520',
              'label': 'energie-waermequelle-heizung',
              'value': 'energie-waermequelle-heizung-7520',
            }),
            'energie-waermequelle-warmwasser': dict({
              'displayValue': '7530',
              'label': 'energie-waermequelle-warmwasser',
              'value': 'energie-waermequelle-warmwasser-7530',
            }),
            'ist-ein-warmwasseranschluss-geplant': dict({
              'label': None,
              'value': 'ja',
            }),
            'waermeerzeuger-heizung': dict({
              'displayValue': '7430',
              'label': 'waermeerzeuger-heizung',
              'value': 'waermeerzeuger-heizung-7430',
            }),
            'waermeerzeuger-warmwasser': dict({
              'displayValue': '7630',
              'label': 'waermeerzeuger-warmwasser',
              'value': 'waermeerzeuger-warmwasser-7630',
            }),
            'weitere-energie-waermequelle-heizung': dict({
              'label': None,
              'value': '7520',
            }),
            'werden-flaechen-beheizt': dict({
              'label': None,
              'value': 'ja',
            }),
          }),
        ]),
      }),
      'genereller-erschliessungsplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Parkierung Gebiete A',
      }),
      'genereller-gestaltungsplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Erhaltungsbereiche Gebiete mit besonderer Wohnqualit√§t',
      }),
      'parzelle': dict({
        'hidden': False,
        'label': None,
        'value': list([
          dict({
            'e-grid-nr': dict({
              'label': None,
              'value': 'CH159786776840',
            }),
            'parzellennummer': dict({
              'label': None,
              'value': '3309',
            }),
          }),
        ]),
      }),
      'zonenplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Wohnzone 3, Keine Gefahrenzone',
      }),
    }),
  })
# ---
# name: test_ech0206_client[2759772.4989033034-1191611.9958049427]
  dict({
    'data': dict({
      'gebaeude-und-anlagen': dict({
        'hidden': False,
        'label': None,
        'value': list([
          dict({
            'amtliche-gebaeudenummer': dict({
              'label': None,
              'value': 'NN',
            }),
            'egid-nr': dict({
              'label': 'egid-nr',
              'value': '191714455',
            }),
            'energie-waermequelle-heizung': dict({
              'displayValue': '7598',
              'label': 'energie-waermequelle-heizung',
              'value': 'energie-waermequelle-heizung-7598',
            }),
            'energie-waermequelle-warmwasser': dict({
              'displayValue': '7598',
              'label': 'energie-waermequelle-warmwasser',
              'value': 'energie-waermequelle-warmwasser-7598',
            }),
            'ist-ein-warmwasseranschluss-geplant': dict({
              'label': None,
              'value': 'ja',
            }),
            'waermeerzeuger-heizung': dict({
              'displayValue': '7410',
              'label': 'waermeerzeuger-heizung',
              'value': 'waermeerzeuger-heizung-7410',
            }),
            'waermeerzeuger-warmwasser': dict({
              'displayValue': '7610',
              'label': 'waermeerzeuger-warmwasser',
              'value': 'waermeerzeuger-warmwasser-7610',
            }),
            'weitere-energie-waermequelle-heizung': dict({
              'label': None,
              'value': '7598',
            }),
            'werden-flaechen-beheizt': dict({
              'label': None,
              'value': 'ja',
            }),
          }),
        ]),
      }),
      'genereller-erschliessungsplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Parkierung Gebiete A',
      }),
      'parzelle': dict({
        'hidden': False,
        'label': None,
        'value': list([
          dict({
            'e-grid-nr': dict({
              'label': None,
              'value': 'CH238677687779',
            }),
            'parzellennummer': dict({
              'label': None,
              'value': '60',
            }),
          }),
        ]),
      }),
      'zonenplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Zone f√ºr √∂ffentliche Bauten und Anlagen, Keine Gefahrenzone',
      }),
    }),
  })
# ---
# name: test_ech0206_client[2760073.7106244904-1191456.8678512666]
  dict({
    'data': dict({
      'gebaeude-und-anlagen': dict({
        'hidden': False,
        'label': None,
        'value': list([
          dict({
            'amtliche-gebaeudenummer': dict({
              'label': None,
              'value': '4-328A',
            }),
            'egid-nr': dict({
              'label': 'egid-nr',
              'value': '9001126',
            }),
            'ist-ein-warmwasseranschluss-geplant': dict({
              'label': None,
              'value': 'nein',
            }),
            'werden-flaechen-beheizt': dict({
              'label': None,
              'value': 'nein',
            }),
          }),
        ]),
      }),
      'genereller-erschliessungsplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Parkierung Gebiete A',
      }),
      'genereller-gestaltungsplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Erhaltungsbereiche Gebiete mit besonderer Wohnqualit√§t',
      }),
      'parzelle': dict({
        'hidden': False,
        'label': None,
        'value': list([
          dict({
            'e-grid-nr': dict({
              'label': None,
              'value': 'CH807768798612',
            }),
            'parzellennummer': dict({
              'label': None,
              'value': '251',
            }),
          }),
          dict({
            'baurecht-nummer': dict({
              'label': None,
              'value': '14463',
            }),
            'e-grid-nr': dict({
              'label': None,
              'value': 'CH296077959995',
            }),
          }),
        ]),
      }),
      'zonenplan': dict({
        'hidden': False,
        'label': None,
        'value': 'Wohnzone 2, Keine Gefahrenzone',
      }),
    }),
  })
# ---
